### PR TITLE
IFontAtlas: font atlas per plugin

### PR DESCRIPTION
**Pull request by Soreepeong**
## Dependency PRs
Should reset to master and then cherry pick the last commit of this PR after below PR merge.

* #1535
* #1536
* #1537
* #1538

## Effective changes
* Each plugin now gets their own font atlas.
    * If a plugin calls `UiBuilder.RebuildFonts()`, then only their fonts will be rebuild. Other plugins and Dalamud's fonts will stay still.
    * With `UiBuilder.CreateFontAtlas(FontAtlasAutoRebuildMode.None/OnNewFrame/Async)`, a plugin can make multiple atlases, should the plugin author decides to make fonts un/loadable without affecting other already loaded fonts. They can also re/load fonts in the background.
    * With `UiBuilder.NewFontHandle`, a plugin can effortlessly make a wrapper for `ImFontPtr` and enjoy automated asynchronous font atlas building, along with not having to manage `BuildFonts` callbacks.
* `ImFontConfigPtr` can be not used, and the safer and easier `SafeFontConfig` can be used instead.
* From `BuildFonts` callbacks and `NewDelegateFontHandle(FontAtlasBuildStepDelegate)` callbacks, `toolkit.Add...` can be used to add fonts in a safer and easier way.
* `ChangelogWindow`, `TitleScreenMenuWindow`, and `SettingsTabAbout` uses asynchronous font loads and updates.

## Code changes
* `Dalamud.Interface.ManagedFontAtlas` have been added.
    * `DalamudFontAtlas` deals with building ImGui font atlas using native function.
    * Everything inside `FontHandle` namespace manages refcounting/handling managing and respective font build procedures.
    * On adding game pre-baked built-in fonts, if no modifications such as italic and bold is done, they will skip ImGui rectpacking and use the game's built in textures divided into 4 for each channel for each texture.
* `GameFonts` lost a lot of unused functions.
    * ImFont modifying capabilities have been moved to `DalamudFontAtlas`.
    * `GameFontStyle` has a typo fix, and better `GetRecommendedFamilyAndSize`.
    * `GameFontHandle` is now a wrapper for `IFontHandle`, for ABI compatibility.
* `InterfaceManager` lost a lot of font features.
    * It now creates `DalamudFontAtlas` to delegate the job of generating font atlas.
* `ImGuiHelpers` have new functions.
    * `AdjustGlyphMetrics`: `UnscaleFont` but now with rounding, and in better categorized class.
    * `NewFontAtlasPtrScoped`, and `NewFontGlyphRangeBuilderPtrScoped` will allocate new native ImGui objects, and wrap it using ImGui.NET wrappers, while returning `IDisposable`s that can be used to deallocate them after use.
    * `BuildRangesScoped` will turn `ImFontGlyphRangesBuilderPtr` into something that can be used with `ImFontConfig.GlyphRanges`.
    * `Mark4KPageUsedAfterGlyphAdd` is just something that needs to be called after adding a glyph to `ImFont.Glyphs`.
* `UiBuilder` now uses its own private font atlas, instead of shared instance at `InterfaceManager`.

### Class definition
```cs
private IFontAtlas fontAtlas;
private IFontHandle font1;
private IFontHandle font2;
private IFontHandle font3;
private IFontHandle font4;
private IFontHandle font5;
```

### Constructor
```cs
this.fontAtlas = pluginInterface.UiBuilder.CreateFontAtlas(FontAtlasAutoRebuildMode.Async);
this.font1 = this.fontAtlas.NewDelegateFontHandle(
    e => e.OnPreBuild(tk => tk.AddDalamudDefaultFont(12)));
this.font2 = this.fontAtlas.NewDelegateFontHandle(
    e => e.OnPreBuild(tk => tk.AddFontAwesomeIconFont(new() { SizePx = 12 })));
this.font3 = this.fontAtlas.NewDelegateFontHandle(
    e => e.OnPreBuild(
        tk => tk.AddDalamudAssetFont(
            DalamudAssetFont.InconsolataRegular,
            new() { SizePx = 12 })));
this.font4 = this.fontAtlas.NewDelegateFontHandle(
    e => e.OnPreBuild(tk => tk.AddDalamudDefaultFont(12, "a".ToGlyphRange())));
this.font5 = this.fontAtlas.NewGameFontHandle(new(GameFontFamilyAndSize.MiedingerMid18));
```

### Draw
```cs
using (this.font1.Push())
    ImGui.TextUnformatted("Font 1");
using (this.font2.Push())
    ImGui.TextUnformatted("Font 2");
using (this.font3.Push())
    ImGui.TextUnformatted("Font 3");
using (this.font4.Push())
    ImGui.TextUnformatted("Font 4");
using (this.font5.Push())
    ImGui.TextUnformatted("Font 5");
```

### CODE REVIEW

**Comment by Soreepeong** [comment]
Checking again after rebasing to the master, to look for anything that stopped working

**Comment by Soreepeong** [comment]
Alright, plugins that utilize extra fonts are tested to be working, regardless of which default font configuration is being used.

**Review by goaaats — COMMENTED** [review]
Just some small stuff to clarify quickly. Looks good in general - I'm split about adding a TrueType font parser to Dalamud, but I assume there's a good reason for why there was no better/easier way that would reuse a pre-existing parser, probably interop?
I don't pretend to have gotten into any of the math in here, but I assume it works :)

**Comment by Soreepeong** [comment]
On TrueType: Typography package has necessary tables defined as internal structs, FreeType(/SharpType) only got supports for necessary tables for rasterization and a few utilities that cannot help us, DirectWrite and stb_truetype only support fetching distance given two glyphs, and so on; font libraries aren't giving us opportunities to extract glyph distances as a whole.

File: Dalamud/Interface/Utility/ImGuiHelpers.cs [hunk 1]
```diff
@@ -406,6 +404,130 @@ public static void CenteredText(string text)
...
+                return memory;
+        }
+    }
+
+    /// <summary>
+    /// Mark 4K page as used, after adding a codepoint to a font.
+    /// </summary>
+    /// <param name="font">The font.</param>
+    /// <param name="codepoint">The codepoint.</param>
+    public static unsafe void Mark4KPageUsedAfterGlyphAdd(this ImFontPtr font, ushort codepoint)
```
[the remark this answers is not in the document]
Comment (Soreepeong) [reply Dalamud/Interface/Utility/ImGuiHelpers.cs:444]: If they decide to add glyphs, then this action will have to be done (for the sake of consistency of how it's done from ImGui native code and from C#.) Didn't quite find what happens differently if this call is missing though. Then again, I don't think any plugin would do that, and as this function is trivial to do it manually, turning this `internal` probably be better.

**Review by goaaats — APPROVED** [review]

**Pull request merged**